Integration with Bitbucket

chandrasekharpanigrahy opened the pull request:
## Summary

<!-- Provide a general summary description of your changes -->

## Details

<!-- Describe your changes in detail -->

## Context

<!-- Why is this change required? What problem does it solve? -->
<!-- Which environnment is/are impacted ? -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I've added tests for my code.
- [ ] Documentation has been updated accordingly to this PR


## Related issue : 
<!-- If it fixes an open issue, please link to the issue here. -->

vincent-fuchs commented on ci-droid-tasks-consumer-bitbucket-adapter/src/test/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/model/PullRequestTest.java:
```diff
@@ -0,0 +1,11 @@
+class PullRequestTest {
+
+    @Test
+    void toStandardPullRequest() {
```
empty test - to remove ?

chandrasekharpanigrahy replied:
implemented test

vincent-fuchs commented:
Thanks a lot for your contribution !!

I've started reviewing it, and one first thing that I think is missing (there may be others), is an integration test showing/confirming that we're able to process the Bitbucket events that we receive. 

If you look at  https://github.com/societe-generale/ci-droid-tasks-consumer/tree/master/ci-droid-tasks-consumer-gitlab-adapter/src/test/resources or https://github.com/societe-generale/ci-droid-tasks-consumer/tree/master/ci-droid-tasks-consumer-github-adapter/src/test/resources , you can see that we use "real" payloads sent by GitLab / GitHub and use them to make sure that we are able to read / convert / process the incoming SourceControlEvents. 

This types of test are really useful when doing some refactoring. I don't see such files being committed for Bitbucket. an integration test using such files and covering BitBucketSourceControlEventController would be required. 

Thanks !

chandrasekharpanigrahy commented:
Thank you Vincent for your quick feedback and starting review. I have implemented Integration test case with example payloads from bitbucket documentation(earlier had implemented with my test data). added missing test case for BitBucketSourceControlEventController. please let me know if I have missed something on this review comment implementation.

vincent-fuchs commented:
there's a problem with the way you use the configured login/password : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/test/resources/application-test.yml#L5-L6 

they are supposed to be used by CI-droid to perform actions like automated rebase. that's why they are used in the Rebaser config : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-autoconfigure/src/main/java/com/societegenerale/cidroid/tasks/consumer/autoconfigure/CiDroidTasksConsumerAutoConfiguration.java#L139-L145 

However, you use the `login ` as a `userSlug  ` to fetch the current user : 

https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L31

https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L122

but the goal of implementations of `SourceControlBulkActionsPerformer ` is to perform actions on behalf of a user who submits a payload.  when the user submits the payload, it contains enough information for the `SourceControlBulkActionsPerformer ` to find him/her in the system, then perform the actions on his/her behalf :

https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/test/resources/incomingSimpleReplaceActionCommand.json#L2-L4 

in your case, I think the gitLogin provided may be the userSlug that needs to be used in https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L121.

the general problem here is that it's not managed the same way from one provider to the other : 
- gitHub and gitLab uses a userToken : 
    - https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-gitlab-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/gitlab/RemoteForGitLabBulkActions.java#L238
    - https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-github-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/github/FeignRemoteForGitHubBulkActions.java#L105
- azureDevops requires to go through the user email, as explained in the comment : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-azure-devops-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/azuredevops/RemoteForAzureDevopsBulkActions.java#L421-L429

So how can we do the same with BitBucket ? 
- if BitBucket allows to retrieve the current user through the userToken then you can do something very similar to gitHub/gitLab.
- if not,  we may need to change the signature of `fetchCurrentUser `:  https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-services/src/main/java/com/societegenerale/cidroid/tasks/consumer/services/SourceControlBulkActionsPerformer.java#L58-L65 
we would add a parameter to pass the gitLogin from the payload : only BitBucket implementation would use it, the others would ignore it.

chandrasekharpanigrahy commented:
> there's a problem with the way you use the configured login/password : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/test/resources/application-test.yml#L5-L6
> 
> they are supposed to be used by CI-droid to perform actions like automated rebase. that's why they are used in the Rebaser config : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-autoconfigure/src/main/java/com/societegenerale/cidroid/tasks/consumer/autoconfigure/CiDroidTasksConsumerAutoConfiguration.java#L139-L145
> 
> However, you use the `login ` as a `userSlug ` to fetch the current user :
> 
> https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L31
> 
> https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L122
> 
> but the goal of implementations of `SourceControlBulkActionsPerformer ` is to perform actions on behalf of a user who submits a payload. when the user submits the payload, it contains enough information for the `SourceControlBulkActionsPerformer ` to find him/her in the system, then perform the actions on his/her behalf :
> 
> https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/test/resources/incomingSimpleReplaceActionCommand.json#L2-L4
> 
> in your case, I think the gitLogin provided may be the userSlug that needs to be used in https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-bitbucket-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/bitbucket/RemoteForBitbucketBulkActionsWrapper.java#L121.
> 
> the general problem here is that it's not managed the same way from one provider to the other :
> 
> * gitHub and gitLab uses a userToken :
>   
>   * https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-gitlab-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/gitlab/RemoteForGitLabBulkActions.java#L238
>   * https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-github-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/github/FeignRemoteForGitHubBulkActions.java#L105
> * azureDevops requires to go through the user email, as explained in the comment : https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-azure-devops-adapter/src/main/java/com/societegenerale/cidroid/tasks/consumer/infrastructure/azuredevops/RemoteForAzureDevopsBulkActions.java#L421-L429
> 
> So how can we do the same with BitBucket ?
> 
> * if BitBucket allows to retrieve the current user through the userToken then you can do something very similar to gitHub/gitLab.
> * if not,  we may need to change the signature of `fetchCurrentUser `:  https://github.com/chandrasekharpanigrahy/ci-droid-tasks-consumer/blob/master/ci-droid-tasks-consumer-services/src/main/java/com/societegenerale/cidroid/tasks/consumer/services/SourceControlBulkActionsPerformer.java#L58-L65
>   we would add a parameter to pass the gitLogin from the payload : only BitBucket implementation would use it, the others would ignore it.

chandrasekharpanigrahy commented:
initially I have implemented as per your second suggestion. But reverted (https://github.com/societe-generale/ci-droid-tasks-consumer/pull/97/commits/ae8443856e1de622992f476dd62b3441fc337293) as it was changing service(ci-droid-tasks-consumer-services) code also.(thought of avoiding changing the service module). 
as you have suggested, now I am using gitLogin from the payload . (Bitbucket API needs user id to return user information, it does not return current user based on email or access token)

vincent-fuchs commented:
Thanks a lot for the contribution, it's merged now ! I will make a release very soon
